### DISCUSSION TITLE
how to pass directory and file name as varaible into file reader source

### DISCUSSION THREAD

**Discussion by aungktun** [nextgenhealthcare/connect#5237] [Q&A] [answered]
I want to read sftp for a particular file under particular directory but that meta info is going to be available from db reader from different channel. I want to pass that into sftp file reader channel to pickup file and do something with it in destination.

How do I pass variable into file reader channel for directory and file name pattern as input variable?

**Comment by aungktun** [comment]
I found alternative answer here. 

https://forums.mirthproject.io/forum/mirth-connect/support/15586-sftp-connection-within-javascript



File Reader as source may not work.

**Comment by siddharthsharma9258** [comment, marked as answer]
you want to use Source Map. 
Typically its just extracting a variable from one channel and sending it to the another channel, and you can capture it at source. 
Read about it.

**Reply by pacmano1** [reply]
@siddharthsharma9258, the ask was for an SFTP poller that does that I think.

**Reply by siddharthsharma9258** [reply]
Well.. this is what OP said in original post "..that meta info is going to be available from db reader from different channel. I want to pass that into sftp file reader channel to pickup file.."

I think his setup is like DB is going to store the name of the file to pickup from SFTP, and work against it.

**Reply by pacmano1** [reply]
And passing that in a map to an SFTP reader makes the File reader poll the SFTP site how?

**Reply by pacmano1** [reply]
So my poiint is you gotta make the SFTP call in java.

**Reply by siddharthsharma9258** [reply]
> So my poiint is you gotta make the SFTP call in java.

That is a possibility, but its highly technical, with the JSch library. Not sure if OP wants to do it.

**Reply by aungktun** [reply]
"> And passing that in a map to an SFTP reader makes the File reader poll the SFTP site how?"
Yes, how ? By the time file reader/sftp source is run, it is expecting the directory and file name and even if I pass the message to channel, not sure how to get that message from the file reader with sftp as source.

**Reply by pacmano1** [reply]
> "> And passing that in a map to an SFTP reader makes the File reader poll the SFTP site how?" Yes, how ? By the time file reader/sftp source is run, it is expecting the directory and file name and even if I pass the message to channel, not sure how to get that message from the file reader with sftp as source.

It doesn't.  And you can't.  That was my point in my answers.

**Reply by aungktun** [reply]
For now I can use db reader as source for pulling meta data for file name and source directory for sftp and in the destination I will use jsch to read the file and do something with that file being read for downstream processes. Thanks.  Here is the code I tested to get the file from sftp using jsch in one of the destinations.

       //original code from https://forums.mirthproject.io/forum/mirth-connect/support/15586-sftp-connection-within-javascript
	importPackage(com.jcraft.jsch);
	importPackage(java.io);
	importPackage(java.lang);
	importPackage(java.nio.charset);	
	
	// testing vars
	var verbose = false;
	
	var b64_content = null
	
	var fileread = null 
	
	//create object.
	var jsch = new JSch();
	jsch.setConfig('StrictHostKeyChecking','no');
	
	var session = jsch.getSession($('SFTP_USR'),$('SFTP_COM'),22);
	session.setPassword($('SFTP_PWD'));
	session.setTimeout(2000);
	session.connect();
	
	var channel = session.openChannel('sftp');
	
	/*connect*/
	channel.connect();	
	
	var remote_path = '/remotepath/'
	var local_path = '/localpath/'
	channel.cd(remote_path); //Folder on the SFTP server where file will be found
	
	var input_fileName = '112939d5-47f1-414b-99f7-10a85f5f75d3.tif'
	var directoryList = channel.ls(input_fileName);	
	
	if (directoryList.size() > 0 ){
	
		for(var i=0; i < directoryList.size();i++){
			var filename = directoryList.get(i).getFilename();			
			//download file
			channel.get(filename, local_path + filename);			
			//read as bytes
			fileread = FileUtil.readBytes(local_path + filename)
			//
			b64_content = FileUtil.encode(fileread)
			//cleanup downloaded file
			FileUtil.deleteFile(new java.io.File(local_path + filename));
		}
	}

	channelMap.put('fileread',fileread);
	channelMap.put('b64_content',b64_content);
	
	
	//Close everything down
	channel.disconnect();
	session.disconnect();

**Reply by pacmano1** [reply]
You can store it in some map and get it on next poll, but you can't send a message to a file reader to make it poll.  So you have to write the SFTP get yourself in java.

**Reply by pacmano1** [reply]
jinx.   Nice work!

**Reply by aungktun** [reply]
yes, sftp get with jsch is definitely working. Thanks.

**Reply by tonygermano** [reply]
In regard to the suggestion at the top of the thread, the job of a source connector is to dispatch messages to the channel. Routing a message from one channel to another will dispatch the message directly and bypass the downstream source connector (File Reader in this case.)

**Reply by siddharthsharma9258** [reply]
Well, I remember doing that using sourceMap long time ago, following below explanation by Nick.

https://forums.mirthproject.io/forum/mirth-connect/support/16328-sourcemap-to-pass-message-parameter-accross-multiple-channels?p=92781#post92781

**Reply by tonygermano** [reply]
You can pass a RawMessage that includes a sourceMap to a channel with a File Reader, but that message will bypass the actual File Reader and start processing immediately without any polling as if the message had been created by the File Reader. You can't pass anything **to** a File Reader (the connector, not the channel) in this manner. It is not capable of receiving input in this way or accessing any maps except config, global, and globalChannel to use when polling.  A File Reader generates its own messages and sourceMaps on the poll schedule.
